# Excercise 4.5

# Shaquille Obomeghie

### Content list 
1. Import Libraries 
2. Import Data product data 
3. Check. for missing data in Product dataframe
4.  Check for Dupicate data in Product Dataframe


# 1. Import Libraries

In [2]:
#importing libaries 
import pandas as pd
import numpy as np
import os

# 2. Importing Product Data

In [3]:
path = r'/Users/shaquillelee/Documents/Instacart Basket Analysis'

In [4]:
#Importing Product data frame
df_prods = pd.read_csv(os.path.join(path, 'Data','Original Data', 'products.csv'))

In [5]:
df_prods.shape

(49693, 5)

# 3. Check. for missing data in Product dataframe

In [6]:
# Find missing value for df_prods
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [7]:
# Missing Data found in product name column. Isolate column to see the missing product names 
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [8]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [9]:
df_nan.describe()

product_id    aisle_id  department_id     prices
count     16.000000   16.000000      16.000000  16.000000
mean    6684.000000   89.937500      10.937500  13.012500
std    12836.665242   33.731229       4.639953   3.881731
min       34.000000   26.000000       1.000000   1.200000
25%      459.250000   70.750000       7.750000  12.175000
50%     2413.000000   98.500000      11.500000  13.650000
75%     3872.750000  120.000000      14.500000  14.425000
max    40440.000000  126.000000      16.000000  20.900000

In [10]:
df_prods.shape

(49693, 5)

In [11]:
# CLean missing data. from prduct name column 
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [12]:
df_prods_clean.shape

(49677, 5)

# 4. Check for Dupicate data in Product Dataframe

In [13]:
# Check. for duplicate data 
df_dups = df_prods_clean[df_prods_clean.duplicated()]


In [14]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [15]:
# Clean Duplicate data from proudct data
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [16]:
df_prods_clean_no_dups.shape

(49672, 5)

# Excercise 4.5


# 5. Import Order Wrangled Data

In [17]:
# import order data
df_ords = pd.read_csv(os.path.join(path, 'Data','Prepared Data', 'orders_wrangled 4.4.csv'),index_col = 0)

In [18]:
df_ords.shape

(3421083, 6)

In [19]:
# 2. Describe the statistics of the df_ords dataframe
df_ords.describe()

order_id  user_identity_number  order_number  orders_day_of_week  \
count  3.421083e+06          3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06          1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05          5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00          1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05          5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06          1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06          1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06          2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

 From my observation it does not seem that theres anthing remotely off about thr statistics of any of the columns in the dataframe . The order_day_of_week has a min of 0 and Max of 6 showing the 7 days of the week. The hour of the day goes from 0 to 23 showing the 24 hours in a day


# 6. Check For Mixed Data in the df_ords Dataframe 

In [20]:
#3. Checking for mixed data in the df_ords dataframe 
for col in df_ords.columns.tolist():  # Loop through each column in the DataFrame
  # Check for inconsistent data types in the column
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis=1)
  
  if len(df_ords[weird]) > 0:  # If there are any rows with inconsistent types
    print(col)  # Print the column name


/var/folders/z4/snghvzgj29jd9tc1ph3523_h0000gn/T/ipykernel_9532/1889894156.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis=1)
/var/folders/z4/snghvzgj29jd9tc1ph3523_h0000gn/T/ipykernel_9532/1889894156.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis=1)
/var/folders/z4/snghvzgj29jd9tc1ph3523_h0000gn/T/ipykernel_9532/1889894156.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis=1)
/var/folders/z4/snghvzgj29jd9tc1ph3523_h0000gn/T/ipykernel_9532/1889894156.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply

There is no duplicate data in our dataframe 

In [21]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3421083 entries, 0 to 3421082
Data columns (total 6 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_identity_number    int64  
 2   order_number            int64  
 3   orders_day_of_week      int64  
 4   order_hour_of_day       int64  
 5   days_since_prior_order  float64
dtypes: float64(1), int64(5)
memory usage: 182.7 MB


# 7. Finding missing values in df_ords dataframe

In [22]:
# 5. Find missing values for df_ords
df_ords.isnull().sum() #We used the Sum() function for the total of missing values 

order_id                       0
user_identity_number           0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

days_since_prior_order have 206209 missing values. There's a posibility they days were not recorded or the information was not available 


In [23]:
# Address missing values 
df_ords_missing = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [24]:
df_ords_missing

order_id  user_identity_number  order_number  orders_day_of_week  \
0         2539329                     1             1                   2   
11        2168274                     2             1                   2   
26        1374495                     3             1                   1   
39        3343014                     4             1                   6   
45        2717275                     5             1                   3   
...           ...                   ...           ...                 ...   
3420930    969311                206205             1                   4   
3420934   3189322                206206             1                   3   
3421002   2166133                206207             1                   6   
3421019   2227043                206208             1                   1   
3421069   3154581                206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

In [25]:
# Filling the missing value with the median '7'
df_ords['days_since_prior_order'].fillna(7, inplace = True)

/var/folders/z4/snghvzgj29jd9tc1ph3523_h0000gn/T/ipykernel_9532/3015464827.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_ords['days_since_prior_order'].fillna(7, inplace = True)


In [26]:
# Median from columns
df_ords.median()

order_id                  1710542.0
user_identity_number       102689.0
order_number                   11.0
orders_day_of_week              3.0
order_hour_of_day              13.0
days_since_prior_order          7.0
dtype: float64

In [27]:
# I have successfully filled in the missing data with the median value 
df_ords.isnull().sum()

order_id                  0
user_identity_number      0
order_number              0
orders_day_of_week        0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

I used the the median to fill in the missing values for the days_since_prior_order. The mean was not used because the value ranges from 0-30 which could cause a huge disaprity in the mean value due to the large gap between the min and max numbers

# 8. Check For duplicate Values in the df_ords dataframe

In [28]:
#7 Run a duplicate check for values in the df_ords data
df_duplicates = df_ords[df_ords.duplicated()]

In [29]:
df_duplicates

Empty DataFrame
Columns: [order_id, user_identity_number, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

There are no duplicate values in our dataframe

# 9. Save updated dataframes 

In [30]:
#Export new data frames
df_ords.to_csv(os.path.join(path, 'Data','Prepared Data', 'orders_data_checked.csv'))
df_prods.to_csv(os.path.join(path, 'Data', 'Prepared Data', 'product_data_checked.csv'))